src/Microsoft.VisualStudio.Services.Agent/ProcessInvoker.cs

```diff
@@ -194,41 +177,132 @@ private void Dispose(bool disposing)
+            while (_outputData.TryDequeue(out outputLine))
+            {
+                outputData.Add(outputLine);
+            }
+
+            _outputProcessEvent.Reset();
+
+            // Write the error lines.
+            if (errorData != null)
+            {
+                foreach (string line in errorData)
+                {
+                    if (line != null && this.ErrorDataReceived != null)
+                    {
+                        this.ErrorDataReceived(this, new ProcessDataReceivedEventArgs(line));
+                    }
+                }
+            }
+
+            // Process the output lines.
+            if (outputData != null)
+            {
+                foreach (string line in outputData)
+                {
+                    if (line != null && this.OutputDataReceived != null)
+                    {
+                        // The line is output from the process that was invoked.
+                        this.OutputDataReceived(this, new ProcessDataReceivedEventArgs(line));
+                    }
+                }
+            }
         }
 
-        private void OnErrorDataReceived(object sender, DataReceivedEventArgs e)
+        private void CancelProcessTree()
         {
-            // at the end of the process, the event fires one last time with null
-            if (e.Data != null)
+            ArgUtil.NotNull(_proc, nameof(_proc));
+
+            try
+            {
+                // TODO: Send Ctrl+C/Break to process group.
+                if (!_proc.HasExited)
+                {
+                    _proc.Kill();
+                }
+            }
+            catch (InvalidOperationException)
             {
-                ErrorDataReceived?.Invoke(sender, e);
+                // InvalidOperationException can occur if process got terminated by itself between 
+                // HasExited and Kill() calls above.
             }
         }
 
-        private void OnOutputDataReceived(object sender, DataReceivedEventArgs e)
+        private void ProcessExitedHandler(object sender, EventArgs e)
         {
-            // at the end of the process, the event fires one last time with null
-            if (e.Data != null)
+            if ((_proc.StartInfo.RedirectStandardError || _proc.StartInfo.RedirectStandardOutput) && _asyncStreamReaderCount != 0)
             {
-                OutputDataReceived?.Invoke(sender, e);
+                _waitingOnStreams = true;
+
+                Task.Run(async delegate
+                {
+                    // Wait 5 seconds and then kill everythin in the Job just in case
+                    await Task.Delay(TimeSpan.FromSeconds(5));
+                    _processExitedCompletionSource.TrySetResult(true);
+                });
+            }
+            else
+            {
+                _processExitedCompletionSource.TrySetResult(true);
             }
         }
+
+        private void ReadStreamAsync(StreamReader reader, ProcessStreamType streamType)
```

Comment:
The method name is misleading. StartReadSteam might be more appropriate.

It seems strange to me that two threads run in a while loop potentially eating a lot of CPU for machines that do not have many cores. Additionally, every time Set is called on the custom reset event class, a thread is queued to set the event.